### ISSUE TITLE
Train Stage 2: Compositional Embeddings for morpheme combination

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#698]
## Goal
Train compositional embeddings that combine root + affix semantics to handle novel word combinations.

### Architecture
- **Size**: 500K params
- **Components**:
  - Prefix embeddings: 16 prefixes × 8 dims = 128 params
  - Suffix embeddings: 32 suffixes × 8 dims = 256 params
  - Root embeddings: Reuse M0.1 (frozen)
  - Combination MLP: (64 + 8 + 8 + 8) → 128 dims ≈ 500K params
- **Purpose**: Predict word semantics from morpheme composition

### Training Objective
Given a word decomposition, predict the word's semantic embedding:
```python
word = "rehundejo"
morphemes = {
    "prefikso": "re",     # again
    "radiko": "hund",     # dog
    "sufikso": "ej",      # place
    "vortspeco": "o"      # noun ending
}

# Predict compositional embedding
predicted_emb = combine(
    embed("re"),      # prefix
    embed("hund"),    # root (frozen from M0.1)
    embed("ej"),      # suffix
    embed("o")        # ending
)

# Loss: MSE to target word embedding
loss = MSE(predicted_emb, target_emb)
```

#### Phase 1: Extract morpheme decompositions from corpus
```bash
python scripts/data/export_compositional_pairs.py \
  --corpus data/corpus/unified_corpus.jsonl \
  --output data/training/compositional_pairs.jsonl
```

**Output format**:
```jsonl
{
  "word": "rehundejo",
  "radiko": "hund",
  "prefikso": "re",
  "sufiksoj": ["ej"],
  "vortspeco": "o",
  "context": "La hundo revenis al sia rehundejo."
}
```

**Requirements**:
- Use AST annotations (already have morpheme decompositions!)
- Filter: only words with full parse success
- Target: ~200K training examples

#### Phase 2: Create composition training pairs
For each word:
1. Get root embedding from M0.1 (frozen)
2. Train affix embeddings + combination function
3. Predict compositional embedding
4. Compare to context-based target

### Implementation Plan

1. **Data Export** (#698 - to be created)
   - [ ] Export compositional pairs from corpus
   - [ ] Split train/val/test (80/10/10)
   - [ ] Validate coverage of affixes

2. **Model Implementation**
   - [ ] Implement `CompositionalEmbedding` class
   - [ ] Add affix embedding tables
   - [ ] Implement combination MLP
   - [ ] Load frozen root embeddings from M0.1

3. **Training Script**
   - [ ] `scripts/train_compositional.sh`
   - [ ] MSE loss on predicted vs target embeddings
   - [ ] Early stopping (patience=3)
   - [ ] Checkpoint rotation

4. **Evaluation**
   - [ ] Composition accuracy on held-out words
   - [ ] Zero-shot test: predict embeddings for words not in training
   - [ ] Qualitative: nearest neighbors for composed words

### Success Criteria
- [ ] Composition accuracy >80% on held-out words
- [ ] Zero-shot accuracy >70% on novel combinations
- [ ] Affix embeddings are interpretable (e.g., mal- reverses meaning)
- [ ] Training script completes without errors
- [ ] Model checkpoints saved to `models/compositional/`

### Dependencies
- **M0.1 Root Embeddings** (#685, #617-620) - MUST be trained first
- Compositional pairs export script (#698)

### Training Pipeline Position
```
M0.1 Root Embeddings (frozen) → [M0.2 Compositional] → M4 Next-Node Predictor
```

### Validation Tests
```python
# Test 1: Affix consistency
assert similarity(embed("mal-"), -embed("")) > 0.7  # mal- reverses meaning

# Test 2: Compositional generalization
assert similarity(
    compose("re", "hund", "ej", "o"),
    embed("rehundejo")
) > 0.8

# Test 3: Zero-shot novel words
novel_word = "malbonvenejo"  # Never seen in training
predicted = compose("mal", "bon", "ven", "ej", "o")
# Should be close to "unwelcome place" semantics
```

### Related Issues
- #685: Root embeddings (M0.1) - dependency
- #692: AST Generation (uses M0.2)
- #615: Compound root decomposition research
- #590: Advanced semantic constraints (future)

### Training Data Statistics (Target)
- Total words in corpus: ~500K unique
- Words with affixes: ~200K (40%)
- Training pairs: ~200K
- Unique prefixes: 16
- Unique suffixes: 32
- Unique roots: ~40K (from M0.1)

### Model Card
After training, create `models/compositional/MODEL_CARD.md`:
- Architecture details
- Training data description
- Evaluation metrics
- Known limitations
- Example compositions

### Integration Points
Once trained, M0.2 is used by:
- M4: Next-Node Predictor (generate novel words)
- Retrieval: Handle unseen word combinations
- Generation: Predict semantics of compounds

## Next Steps After This Issue
1. Train M1 Selectional Preference (#687)
2. Train M2 Reranker (#686)
3. Train M4 Next-Node Predictor (#692)

**Comment by marctjones** [comment]
## Status: Waiting on RootEmbedder (#685)

MorphemeComposer requires frozen RootEmbedder embeddings as input.

### Current Status
- ✅ RootEmbedder improvements integrated (antonym pairs)
- ⏸️ Waiting for corpus rebuild + retraining
- ⏸️ This issue blocked until RootEmbedder frozen_v1.0.pt is ready

### What Happens Next
1. Complete RootEmbedder training (#685)
2. Freeze model → `models/root_embedder/frozen_v1.0.pt`
3. **Then** start MorphemeComposer implementation (this issue)

### Estimated Timeline
- RootEmbedder: 3-4 hours (corpus rebuild + training)
- MorphemeComposer: 1-2 days (after RootEmbedder completes)

Updated #685 with next steps.

**Comment by marctjones** [comment]
Closing as part of a 2026-09-08 backlog cleanup (see PLAN.md, DESIGN.md, CLAUDE.md's merge gate).

This issue predates the merge gate (adopted 2026-07-13) and the AST-native orchestrator rewrite. It describes work against the pre-orchestrator RAG pipeline or the learned model stack (M1/M2 embeddings, compositional embeddings, entity classifiers, neural rerankers) that was pruned from the repo -- klareco/embeddings/, klareco/models/, and klareco/summarization/ no longer exist. It has no milestone and was never benchmarked, matching exactly the failure mode CLAUDE.md documents (25 capability issues filed 2026-05-26, 8 implemented, none benchmarked).

Labeled archived-2026-09 so this batch is recoverable: gh issue list --state closed --label archived-2026-09. If this capability is still wanted, re-file it against the current architecture naming the benchmark metric and test the merge gate requires.
